admin/aws.py

```diff
@@ -471,7 +480,45 @@ def _perform_upload_s3_key(self, dispatcher, intent):
         """
         bucket = self.s3_buckets[intent.target_bucket]
         with intent.file.open() as source_file:
-            bucket[intent.target_key] = source_file.read()
+            bucket[intent.target_key] = source_file.read(), intent.content_type
+
+    def get_bucket_keys(self, bucket):
+        """
+        Retrieve the keys from a bucket in S3.
+
+        :param str bucket: name of the S3 bucket.
```

Comment:
This and following - `bytes` or `unicode` (almost certainly `bytes`).  `str` makes it too easy to forget about the difference between the two types.  It's probably only correct as the type of things like Python identifiers (eg, the 2nd argument of `getattr` or the elements of the return value of `dir`).